test/gurka/test_multi_level_loki.cc

```diff
@@ -0,0 +1,133 @@
+#include "gurka.h"
+#include "rapidjson/document.h"
+#include <boost/format.hpp>
+#include <boost/optional.hpp>
+#include <gtest/gtest.h>
+
+#if !defined(VALHALLA_SOURCE_DIR)
+#define VALHALLA_SOURCE_DIR
+#endif
+
+using namespace valhalla;
+const std::unordered_map<std::string, std::string> build_config{{}};
+
+namespace {
+struct Waypoint {
+  midgard::PointLL ll;
+  boost::optional<int8_t> preferred_z_level;
+};
+
+std::string ToString(const rapidjson::Document& doc) {
+  rapidjson::StringBuffer sb;
+  rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
+  doc.Accept(writer);
+  return sb.GetString();
+}
+
+std::vector<std::string> GetStepNames(const rapidjson::Document& doc) {
+  std::vector<std::string> names;
+  auto steps = doc["routes"][0]["legs"][0]["steps"].GetArray();
+  names.reserve(steps.Size());
+  for (const auto& step : steps) {
+    names.emplace_back(step["name"].GetString());
+  }
+  return names;
+}
+
+std::string BuildRequest(const std::vector<Waypoint>& waypoints) {
+  rapidjson::Document requestDoc;
+  requestDoc.SetObject();
+  auto& allocator = requestDoc.GetAllocator();
+  rapidjson::Value locations(rapidjson::kArrayType);
+  for (const auto& waypoint : waypoints) {
+    rapidjson::Value p(rapidjson::kObjectType);
+    p.AddMember("lon", waypoint.ll.lng(), allocator);
+    p.AddMember("lat", waypoint.ll.lat(), allocator);
+    if (waypoint.preferred_z_level) {
+      p.AddMember("preferred_z_level", *waypoint.preferred_z_level, allocator);
+      // `preferred_z_level` is only working correctly if radius is provided,
+      // because Z-level filtering is performed as last step in loki and if we don't provide
+      // radius we would have only single candidate on the last step in this case
+      p.AddMember("radius", 1, allocator);
```

Comment:
we support doing this already in the basic gurka functions, you can specify options  by their json paths in an unordered_map. it looks like this:

https://github.com/valhalla/valhalla/blob/master/test/gurka/test_via_waypoints.cc#L77-L81

it uses slashes to do the hierarchical json location, using integers to specify the indices of the array elements